File: scripts/profile_coach_conversion_shopify_export.mjs
import fs from 'node:fs';

const input = '/home/ubuntu/upload/orders_export_1(2).csv';
const output = '/home/ubuntu/coach_conversion_shopify_export_profile.json';

function parseCsv(text) {
  const rows = [];
  let row = [];
  let field = '';
  let quoted = false;
  for (let index = 0; index < text.length; index += 1) {
    const character = text[index];
    if (quoted) {
      if (character === '"' && text[index + 1] === '"') {
        field += '"';
        index += 1;
      } else if (character === '"') {
        quoted = false;
      } else {
        field += character;
      }
    } else if (character === '"') {
      quoted = true;
    } else if (character === ',') {
      row.push(field);
      field = '';
    } else if (character === '\n') {
      row.push(field.replace(/\r$/, ''));
      rows.push(row);
      row = [];
      field = '';
    } else {
      field += character;
    }
  }
  if (field || row.length) {
    row.push(field);
    rows.push(row);
  }
  return rows;
}

function number(value) {
  const parsed = Number.parseFloat(String(value ?? '').trim());
  return Number.isFinite(parsed) ? parsed : 0;
}

const [header, ...rawRows] = parseCsv(fs.readFileSync(input, 'utf8'));
const rows = rawRows.filter((row) => row.length > 1).map((row) => Object.fromEntries(header.map((key, index) => [key, row[index] ?? ''])));
const orderMap = new Map();
const productMap = new Map();

for (const row of rows) {
  const id = row.Id?.trim();
  if (!id) continue;
  const status = row['Financial Status']?.trim().toLowerCase() || 'unknown';
  const source = row.Source?.trim() || 'unknown';
  if (!orderMap.has(id)) {
    orderMap.set(id, {
      status,
      source,
      total: number(row.Total),
      refunded: number(row['Refunded Amount']),
    });
  }
  const key = `${row['Lineitem name']?.trim() || '(unnamed)'}\u0000${row['Lineitem sku']?.trim() || ''}`;
  const current = productMap.get(key) || { name: row['Lineitem name']?.trim() || '(unnamed)', sku: row['Lineitem sku']?.trim() || null, quantity: 0, grossLineValueAfterLineDiscount: 0 };
  const quantity = Math.max(0, Number.parseInt(row['Lineitem quantity'], 10) || 0);
  current.quantity += quantity;
  current.grossLineValueAfterLineDiscount += number(row['Lineitem price']) * quantity - number(row['Lineitem discount']);
  productMap.set(key, current);
}

const byStatus = {};
const bySource = {};
for (const order of orderMap.values()) {
  const status = byStatus[order.status] || { orders: 0, total: 0, refunded: 0 };
  status.orders += 1;
  status.total += order.total;
  status.refunded += order.refunded;
  byStatus[order.status] = status;
  bySource[order.source] = (bySource[order.source] || 0) + 1;
}

const profile = {
  sourceFile: 'orders_export_1(2).csv',
  columns: header,
  rawLineRows: rows.length,
  uniqueOrders: orderMap.size,
  uniqueOrdersByFinancialStatus: Object.fromEntries(Object.entries(byStatus).map(([key, value]) => [key, { ...value, total: value.total.toFixed(2), refunded: value.refunded.toFixed(2) }])),
  uniqueOrdersBySource: bySource,
  products: [...productMap.values()]
    .sort((left, right) => right.quantity - left.quantity || left.name.localeCompare(right.name))
    .map((product) => ({ ...product, grossLineValueAfterLineDiscount: product.grossLineValueAfterLineDiscount.toFixed(2) })),
};

const piiTerms = /email|name|address|phone|company|city|zip|province|country|customer|billing|shipping/i;
const categoricalFields = header.filter((column) => !piiTerms.test(column));
const attributionFieldSummary = {};
for (const column of categoricalFields) {
  const values = new Map();
  let nonEmptyRows = 0;
  for (const row of rows) {
    const value = String(row[column] ?? '').trim();
    if (!value) continue;
    nonEmptyRows += 1;
    if (values.size <= 50) values.set(value, (values.get(value) || 0) + 1);
  }
  if (nonEmptyRows > 0 && values.size <= 50) {
    attributionFieldSummary[column] = {
      nonEmptyRows,
      values: Object.fromEntries([...values.entries()].sort(([a], [b]) => a.localeCompare(b))),
    };
  }
}
profile.nonPiiCategoricalFieldSummary = attributionFieldSummary;

fs.writeFileSync(output, JSON.stringify(profile, null, 2));
console.log(output);

File: scripts/profile_coach_conversion_shopify_export.py
import csv
import json
from collections import Counter, defaultdict
from decimal import Decimal, InvalidOperation
from pathlib import Path

INPUT = Path('/home/ubuntu/upload/orders_export_1(2).csv')
OUTPUT = Path('/home/ubuntu/coach_conversion_shopify_export_profile.json')


def money(value: str) -> Decimal:
    try:
        return Decimal((value or '0').strip() or '0')
    except InvalidOperation:
        return Decimal('0')


def main() -> None:
    rows = list(csv.DictReader(INPUT.open(newline='', encoding='utf-8-sig')))
    orders: dict[str, dict] = {}
    products: Counter[tuple[str, str]] = Counter()
    product_revenue: defaultdict[tuple[str, str], Decimal] = defaultdict(Decimal)
    status_counts: Counter[str] = Counter()
    source_counts: Counter[str] = Counter()

    for row in rows:
        order_id = (row.get('Id') or '').strip()
        if not order_id:
            continue
        status = (row.get('Financial Status') or '').strip().lower()
        status_counts[status] += 1
        source_counts[(row.get('Source') or '').strip() or 'unknown'] += 1
        product_key = ((row.get('Lineitem name') or '').strip(), (row.get('Lineitem sku') or '').strip())
        quantity = int((row.get('Lineitem quantity') or '0').strip() or '0')
        products[product_key] += quantity
        product_revenue[product_key] += money(row.get('Lineitem price')) * quantity - money(row.get('Lineitem discount'))
        if order_id not in orders:
            orders[order_id] = {
                'status': status,
                'total': money(row.get('Total')),
                'refunded': money(row.get('Refunded Amount')),
                'source': (row.get('Source') or '').strip() or 'unknown',
                'line_items': 0,
            }
        orders[order_id]['line_items'] += 1

    order_statuses = Counter(order['status'] for order in orders.values())
    order_sources = Counter(order['source'] for order in orders.values())
    output = {
        'source_file': INPUT.name,
        'raw_line_rows': len(rows),
        'unique_orders': len(orders),
        'unique_orders_by_financial_status': dict(order_statuses),
        'unique_orders_by_source': dict(order_sources),
        'line_rows_by_financial_status': dict(status_counts),
        'line_rows_by_source': dict(source_counts),
        'total_order_value_by_financial_status': {
            status: f"{sum((order['total'] for order in orders.values() if order['status'] == status), Decimal('0')):.2f}"
            for status in sorted(order_statuses)
        },
        'total_refunded_by_financial_status': {
            status: f"{sum((order['refunded'] for order in orders.values() if order['status'] == status), Decimal('0')):.2f}"
            for status in sorted(order_statuses)
        },
        'products': [
            {
                'name': name,
                'sku': sku or None,
                'quantity': products[(name, sku)],
                'gross_line_value_after_line_discount': f"{product_revenue[(name, sku)]:.2f}",
            }
            for name, sku in sorted(products, key=lambda key: (-products[key], key[0].lower(), key[1].lower()))
        ],
    }
    OUTPUT.write_text(json.dumps(output, indent=2), encoding='utf-8')
    print(OUTPUT)


if __name__ == '__main__':
    main()
